homeassistant/components/lektrico/sensor.py

```diff
@@ -0,0 +1,294 @@
+"""Support for Lektrico charging station sensors."""
+from __future__ import annotations
+
+from dataclasses import dataclass
+from datetime import datetime, timedelta
+import logging
+from typing import Any
+
+from lektricowifi import lektricowifi
+
+from homeassistant.components.sensor import (
+    DEVICE_CLASS_CURRENT,
+    DEVICE_CLASS_ENERGY,
+    DEVICE_CLASS_ILLUMINANCE,
+    DEVICE_CLASS_POWER,
+    DEVICE_CLASS_TEMPERATURE,
+    DEVICE_CLASS_VOLTAGE,
+    STATE_CLASS_MEASUREMENT,
+    STATE_CLASS_TOTAL_INCREASING,
+    SensorEntity,
+    SensorEntityDescription,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import (
+    ATTR_IDENTIFIERS,
+    ATTR_MANUFACTURER,
+    ATTR_MODEL,
+    ATTR_NAME,
+    ATTR_SW_VERSION,
+    CONF_FRIENDLY_NAME,
+    ELECTRIC_CURRENT_AMPERE,
+    ELECTRIC_POTENTIAL_VOLT,
+    ENERGY_KILO_WATT_HOUR,
+    POWER_KILO_WATT,
+    TEMP_CELSIUS,
+    TIME_SECONDS,
+)
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryNotReady
+from homeassistant.helpers import device_registry as dr, entity_registry as er
+from homeassistant.helpers.entity import DeviceInfo
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.update_coordinator import (
+    CoordinatorEntity,
+    DataUpdateCoordinator,
+)
+
+from .const import DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+PARALLEL_UPDATES = 1
+SCAN_INTERVAL = timedelta(seconds=10)
+
+
+@dataclass
+class LektricoSensorEntityDescription(SensorEntityDescription):
+    """A class that describes the Lektrico sensor entities."""
+
+    unit_fn: str | None = None
+    value_fn: Any | None = None
```

Comment:
Please don't use `Any`, instead, type it properly.

Reply:
I tried to change:
`value_fn: Any | None = None`
to
`value_fn: str | int | float | bool | None = None`
and removed any Any, but mypy shows errors at every LektricoSensorEntityDescription at the line:
`value_fn=lambda x: x.charger_state,`
The errors look like this:
`
homeassistant/components/lektrico/sensor.py:48: error: Argument "value_fn" to "LektricoSensorEntityDescription" has incompatible type "Callable[[Any], Any]"; expected "Union[str, int, float, bool, None]"  [arg-type]
homeassistant/components/lektrico/sensor.py:54: error: Argument "value_fn" to "LektricoSensorEntityDescription" has incompatible type "Callable[[Any], Any]"; expected "Union[str, int, float, bool, None]"  [arg-type]
homeassistant/components/lektrico/sensor.py:62: error: Argument "value_fn" to "LektricoSensorEntityDescription" has incompatible type "Callable[[Any], Any]"; expected "Union[str, int, float, bool, None]"  [arg-type]
homeassistant/components/lektrico/sensor.py:70: error: Argument "value_fn" to "LektricoSensorEntityDescription" has incompatible type "Callable[[Any], Any]"; expected "Union[str, int, float, bool, None]"  [arg-type]
homeassistant/components/lektrico/sensor.py:77: error: Argument "value_fn" to "LektricoSensorEntityDescription" has incompatible type "Callable[[Any], Any]"; expected "Union[str, int, float, bool, None]"  [arg-type]
homeassistant/components/lektrico/sensor.py:85: error: Argument "value_fn" to "LektricoSensorEntityDescription" has incompatible type "Callable[[Any], Any]"; expected "Union[str, int, float, bool, None]"  [arg-type]
homeassistant/components/lektrico/sensor.py:93: error: Argument "value_fn" to "LektricoSensorEntityDescription" has incompatible type "Callable[[Any], Any]"; expected "Union[str, int, float, bool, None]"  [arg-type]
homeassistant/components/lektrico/sensor.py:100: error: Argument "value_fn" to "LektricoSensorEntityDescription" has incompatible type "Callable[[Any], Any]"; expected "Union[str, int, float, bool, None]"  [arg-type]
homeassistant/components/lektrico/sensor.py:107: error: Argument "value_fn" to "LektricoSensorEntityDescription" has incompatible type "Callable[[Any], Any]"; expected "Union[str, int, float, bool, None]"  [arg-type]
homeassistant/components/lektrico/sensor.py:114: error: Argument "value_fn" to "LektricoSensorEntityDescription" has incompatible type "Callable[[Any], Any]"; expected "Union[str, int, float, bool, None]"  [arg-type]
homeassistant/components/lektrico/sensor.py:122: error: Argument "value_fn" to "LektricoSensorEntityDescription" has incompatible type "Callable[[Any], Any]"; expected "Union[str, int, float, bool, None]"  [arg-type]
homeassistant/components/lektrico/sensor.py:127: error: Argument "value_fn" to "LektricoSensorEntityDescription" has incompatible type "Callable[[Any], Any]"; expected "Union[str, int, float, bool, None]"  [arg-type]
homeassistant/components/lektrico/sensor.py:184: error: Returning Any from function declared to return "Union[float, str, int, bool, None]"  [no-any-return]
homeassistant/components/lektrico/sensor.py:184: error: "str" not callable  [operator]
homeassistant/components/lektrico/sensor.py:184: error: "float" not callable  [operator]`

I tried to put types in but I can not make it work. :(
Can you please show me how to change them?

Reply:
For a start try `Callable[[Any], Any]`. `value_fn` is a function call, not an int, float, etc.

To improve further, the first `Any` are the parameters to `value_fn`. So to have tighter typing than `Any` you could change `[Any]` to `[str | int | float | bool]`.

The second `Any` is the return value from the `value_fn`.

Reply:
Hi, thank you for your answer.
I extended the sensors as Frenk suggested above and I succeeded to remove Any.
Thank you.